google-cloud-firestore/lib/google/cloud/firestore/collection.rb

```diff
@@ -0,0 +1,663 @@
+            document_path ||= random_document_id
+
+            ensure_context!
+            context.doc "#{collection_path}/#{document_path}"
+          end
+          alias_method :document, :doc
+
+          ##
+          # Create a document with random document identifier.
+          #
+          # The batch will fail if the document already exists.
+          #
+          # @param [Hash] data The document's fields and values. Optional.
+          #
+          # @return [Document::Reference] A created document.
+          #
+          # @example Create a document with a random ID:
+          #   require "google/cloud/firestore"
+          #
+          #   firestore = Google::Cloud::Firestore.new
+          #
+          #   # Get a collection reference
+          #   cities_col = firestore.col "cities"
+          #
+          #   # Get a document reference without data
+          #   random_ref = cities_col.add
+          #
+          #   # The document ID is randomly generated
+          #   random_ref.document_id #=> "RANDOMID123XYZ"
+          #
+          # @example Create a document with data:
+          #   require "google/cloud/firestore"
+          #
+          #   firestore = Google::Cloud::Firestore.new
+          #
+          #   # Get a collection reference
+          #   cities_col = firestore.col "cities"
+          #
+          #   # Get a document reference with data
+          #   random_ref = cities_col.add({ name: "New York City" })
+          #
+          #   # The document ID is randomly generated
+          #   random_ref.document_id #=> "RANDOMID123XYZ"
+          #
+          def add data = nil
+            data ||= {}
+            doc.tap { |d| d.create data }
+          end
+
+          ##
+          # Retrieves a list of documents with data.
+          #
+          # @param [String, Document::Reference] docs One or more strings
+          #   representing the path of the document, or document reference
+          #   objects.
+          # @param [Array<String|Symbol>, String|Symbol] mask A list of field
+          #   paths to filter the returned document data by.
+          #
+          # @yield [documents] The block for accessing the documents with data.
+          # @yieldparam [Document::Snapshot] document A document with data.
+          #
+          # @return [Enumerator<Document::Snapshot>] documents with data list.
+          #
+          # @example
+          #   require "google/cloud/firestore"
+          #
+          #   firestore = Google::Cloud::Firestore.new
+          #
+          #   # Get a collection reference
+          #   cities_col = firestore.col "cities"
+          #
+          #   # Get and print city documents
+          #   cities = ["cities/NYC", "cities/SF", "cities/LA"]
+          #   cities_col.get_all(cities).each do |city|
+          #     puts "#{city.document_id} has #{city[:population]} residents."
+          #   end
+          #
+          def get_all *docs, mask: nil, &block
```

Comment:
This should sit on Firestore and accept documents in any collection.

Reply:
Yes, and it does. `Database#get_all` accepts documents of any collection. `Collection::Reference#get_all` also exists as a convenience.

Also, users can call `Collection::Reference#get_all` and pass document reference objects that belong to other collections and the operation will return the appropriate document snapshot.

Reply:
The last behavior you described seems confusing. Is there a reason we need to have this method exposed in two places? We would ideally want all clients to be the same, as this makes porting code from one platform to the others much easier. With this addition, porting from Ruby to other languages is no longer straightforward.

Reply:
I don't understand. The requested API is still present, we are simply adding additional conveniences that we think Ruby developers would appreciate. Is this not okay to do?

Reply:
In my mind (and I think I speak for all of us on the Firestore client side), any API addition on any platform is only justified if it is required to make the client idiomatic for its environment. So if Ruby users expect these methods to exist on every class (sorry, I am exaggerating here, but I also have never really used Ruby before), then this is OK. I do suspect that this is not the case though. 

We have a very clearly defined API (which went to months of API review process and refinement) and we are using the same API in all of our Mobile Clients (Android, Web, iOS, C++) and the six released Server SDKs. For anyone that is familiar with any of the the supported languages, it is currently extremely straightforward to port code from one platform to another. While one language might use Promises and another one may use Tasks, the existing API differences can be explained by the underlying platform. If the Ruby client diverges from this API, Firestore Ruby will no longer adhere to the same API standards and our shared customers will likely start to expect certain API features from the existing implementations that we have explicitly decided not to implemented.

We are still open to suggestions when it comes to our API surface, but we value consistency a lot and would like to ensure that every user of Firestore gets the exact same experience, no matter what language they choose to use. If Ruby supports API endpoints that are not available in other SDKs, then that is no longer the case.

As one data point, here is users asking for Query.select() support on our mobile SDKs because it exists in Node: https://github.com/firebase/firebase-js-sdk/issues/212

Reply:
Well, I think it is a mistake not to take the accepted norms and idioms of each language and environment into account, but I don't think I will be able to convince you that these affordances and enhancements are in addition to the clearly defined api, and not a divergence from it. I'll close this PR and open a separate PR that has all these conveniences removed.